File: Baekjoon/Class3/07.1003 피보나치 함수.py
#반복 동적 프로그래밍
import sys
input = sys.stdin.readline
write = sys.stdout.write

FIB_MAX = 40

#타뷸레이션
zero_count = [0] * (FIB_MAX + 2)
zero_count[0] = 1
for i in range(2, FIB_MAX + 2):
    zero_count[i] = zero_count[i - 1] + zero_count[i - 2]

t = int(input())
for _ in range(t):
    n = int(input())
    write(f"{zero_count[n]} {zero_count[n + 1]}\n")

''' 재귀 동적 프로그래밍
import sys
input = sys.stdin.readline
write = sys.stdout.write

FIB_MAX = 40

#메모이제이션
zero_count = [0] * (FIB_MAX + 2)
zero_count[0] = 1

def count_fib(n):
    if n < 2: return
    if zero_count[n] != 0: return
    count_fib(n - 1)
    count_fib(n - 2)
    zero_count[n] = zero_count[n - 1] + zero_count[n - 2]

def main():
    t = int(input())
    for _ in range(t):
        n = int(input())
        count_fib(n + 1)
        write(f"{zero_count[n]} {zero_count[n + 1]}\n")

main()
'''

File: Baekjoon/Class3/07.1003 피보나치 함수.js
// 반복 동적 프로그래밍
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n").map(Number);

const FIB_MAX = 40;

// 타뷸레이션
const zeroCount = new Array(FIB_MAX + 2).fill(0);
zeroCount[0] = 1;
for(let i = 2; i < FIB_MAX + 2; i++)
    zeroCount[i] = zeroCount[i - 1] + zeroCount[i - 2];

const t = input[0];
const out = [];
for(let i = 1; i <= t; i++) {
    const n = input[i];
    out.push(`${zeroCount[n]} ${zeroCount[n + 1]}`);
}
console.log(out.join("\n"));

/* 재귀 동적 프로그래밍
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n").map(Number);

const FIB_MAX = 40;

// 메모이제이션
const zeroCount = new Array(FIB_MAX + 2).fill(0);
zeroCount[0] = 1;

const countFib = (n) => {
    if(n < 2) return;
    if(zeroCount[n] !== 0) return;
    countFib(n - 1);
    countFib(n - 2);
    zeroCount[n] = zeroCount[n - 1] + zeroCount[n - 2];
}

const t = input[0];
const out = [];
for(let i = 1; i <= t; i++) {
    const n = input[i];
    countFib(n + 1);
    out.push(`${zeroCount[n]} ${zeroCount[n + 1]}`);
}
console.log(out.join("\n"));
*/